src/TypeSpec.Extension/Emitter.Csharp/src/lib/confidentLevels.ts

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import {
+    InputDictionaryType,
+    InputListType,
+    InputModelType,
+    InputType,
+    isInputDictionaryType,
+    isInputListType,
+    isInputModelType
+} from "../type/inputType.js";
+
+export function getConfident(
+    type: InputType,
+    visitedModels: Set<InputModelType>
+): boolean {
+    if (isInputModelType(type)) {
+        return getModelConfident(type, visitedModels);
+    } else if (isInputListType(type)) {
+        return getListConfident(type, visitedModels);
+    } else if (isInputDictionaryType(type)) {
+        return getDictionaryConfident(type, visitedModels);
+    }
+
+    return type.IsConfident ?? true; // by default they are confident
+}
+
+function getModelConfident(
+    type: InputModelType,
+    visitedModels: Set<InputModelType>
+): boolean {
+    if (visitedModels.has(type)) return type.IsConfident ?? true; // this means this has been calculated before or being calculated right now, return
+    visitedModels.add(type);
+
+    if (type.IsConfident !== null) return type.IsConfident;
+
+    const isBaseConfident = type.BaseModel
```

Comment:
Just little suggestion: when we meet `false` once, we can directly return `false` and can skip further calculation.

Reply:
Yes I already did this now. I should see it if you refresh